### PR TITLE
Language Server Protocol for TLAPM

### LINKED ISSUE #90: LSP support
This is related to https://github.com/tlaplus/vscode-tlaplus/issues/153.

The current TLAPS/Toolbox integration involves document parsing on the Toolbox side (it uses Sany for that, as I understand). While moving to the VSCode, I would like to push as much logic out of the editor to the backend (TLAPS). The parsing is one aspect. The TLAPS has a parser implementation anyway, and the parser is used by the editor in the scope of the proofs, at least for locating the obligations (or a sub-tree of obligations), folding, and probably proof decomposition functionality.

Instead of implementing a specific protocol for the TLAPS integration in the VSCode, it may be worth trying to implement LSP (Language Server Protocol) internally in the TLAPS. E.g., if started with the `--lsp` option instead of the `--toolbox`, it would behave as an LSP server. It could then receive a document to process, handle commands (e.g., check a step, propose a decomposition, etc.), and output the relevant information.

It looks like there are some LSP building blocks that would be reused, e.g., https://github.com/ocaml/ocaml-lsp.
I hope it should not interfere with the existing functionality as VSCode supports multiple LSP servers for a single document.

What do you think about such a direction?

### PR DESCRIPTION
**Pull request by kape1395**
The changes are over #83 and #91. Closes #90 and #94. The corresponding client part for the VSCode extension is already merged as https://github.com/tlaplus/vscode-tlaplus/pull/307.

  - It invokes the parser directly (as a library function) to analyze the document structure and then invokes `tlapm` as a sub-process.
  - The changes to the `tlapm` are:
      - Accept a file to check from a stdin. This allows the LSP server to check unsaved documents.
      - Accept a file to check as a string when calling the TLAPM parser as a library. Used by the LSP server to locate proof steps.
      - The location tracking for the QED step was broken in the parser. Now that's fixed. Used in the LSP server to show the proof status.
  - The LSP server's part requires OCaml-5.1 because I use Eio for concurrency. I made it optional so the TLAPM can still be compiled in older OCaml versions.

From the functionality perspective:
  - Proof checking can be invoked by pressing `ctrl+g+g` or via the code action (yellow lamp along the cursor position).
  - Proof status is shown by marking document lines in different colors; the hover explanations are also present.
  - Failed proofs are shown as diagnostics.
  - Progress of the proof checking is shown (status bar). The operation can be canceled.
  - If a new proof-checking command is invoked while the old one is still running, the old invocation is canceled automatically.

### CODE REVIEW

**Comment by damiendoligez** [comment]
Now that #83 and #91 are merged, could you rebase this PR? This will make reviewing easier.

**Comment by kape1395** [comment]
Rebased it now.

**Comment by kape1395** [comment]
One of the tests started to fail. That's `test/fast/regression/setEuclid_test.tla`, it fails with
```
File "./setEuclid_test.tla", line 364, characters 9-10:
[ERROR]: Could not prove or check:
           ASSUME NEW CONSTANT Input,
                  InputAssump ==
                    /\ Input \subseteq Nat \ {0}
                    /\ Input # {}
                    /\ IsFiniteSet(Input),
                  NEW CONSTANT defaultInitValue,
                  NEW VARIABLE S,
                  NEW VARIABLE gcd,
                  NEW VARIABLE pc,
                  TypeOK ==
                    /\ /\ S \subseteq Nat \ {0}
                       /\ S # {}
                       /\ IsFiniteSet(S)
                    /\ pc = "Done" => gcd \in Nat \ {0}
                    /\ pc \in {"lbl", "Done"},
                  Inv == /\ TypeOK
                         /\ GCD(S) = GCD(Input)
                         /\ CorrectTermination,
                  Inv ,
                  Next ,
                  CardThm 
           PROVE  Cardinality(S) = 1 => S = {CHOOSE t \in S : TRUE}
File "./setEuclid_test.tla", line 1, character 1 to line 379, character 77:
[ERROR]: 1/104 obligations failed.
There were backend errors processing module `"setEuclid_test"`.
```

From this log, it looks like it started to fail after the `smt-fix` branch was merged.
https://github.com/tlaplus/tlapm/commits/main

Do you have a clue where to look for the reason?

**Comment by muenchnerkindl** [comment]
I'd say the issue is that `CardThm' is not expanded, how come?

**Comment by kape1395** [comment]
Can it happen not because CardThm has not expanded but because some cases for the CHOOSE handling are disabled in https://github.com/tlaplus/tlapm/pull/55/files?

**Comment by muenchnerkindl** [comment]
I don't know about that but the theorem is certainly necessary for proving this step, since none of the backends handle the Cardinality operator natively. So I'd first try to establish why the obligation only shows the name of the theorem, not the assertion (the PM is supposed to expand theorem names).

**Comment by kape1395** [comment]
Sharing some findings. The CardThm is expanded for some provers (zenon, isabelle), e.g.
```
@!!BEGIN
@!!type:obligation
@!!id:98
@!!loc:364:9:364:11
@!!status:being proved
@!!prover:isabelle
@!!meth:auto; time-limit: 150
@!!obl:
ASSUME NEW CONSTANT Input,
       NEW CONSTANT defaultInitValue,
       NEW VARIABLE S,
       NEW VARIABLE gcd,
       NEW VARIABLE pc,
       /\ /\ /\ S \subseteq Nat \ {0}
             /\ S # {}
             /\ IsFiniteSet(S)
          /\ pc = "Done" => gcd \in Nat \ {0}
          /\ pc \in {"lbl", "Done"}
       /\ GCD(S) = GCD(Input)
       /\ CorrectTermination ,
       Next ,
       \A T :
          IsFiniteSet(T)
          => (/\ Cardinality(T) \in Nat
              /\ Cardinality(T) = 0 <=> T = {}
              /\ Cardinality(T) = 1 <=> (\E x \in T : T = {x})) 
PROVE  Cardinality(S) = 1 => S = {CHOOSE t \in S : TRUE}

@!!END
```
But for `smt`, it is not expanded.

Also, the step that fails now succeeds in the "pre `smt-fix` version" even without `CardThm` theorem at all.
```
      <3>1. Cardinality (S) = 1 => S = {CHOOSE t \in S : TRUE}
        BY (*CardThm*) DEF Inv, TypeOK
```
This probably confirms the need for the smt-fix.

**Comment by damiendoligez** [comment]
`CardThm` is not expanded in the error message but it **is** expanded in the obligation sent to the prover.

I have confirmed that the failure is triggered by #55, particularly by the removal of the (unsound) rewrite of
`x = CHOOSE y : P(y)` to `\E z : P(z) => P(x)`.

I found out two things:
- the obligation fails because of "internal timeout", which most probably means that the rewrite engine is now looping.
- the result of the rewrite rule is nowhere to be found in the input to the SMT solver.

I'll continue my investigations.

**Comment by uguryavuz** [comment]
I built the LSP server following [the instructions here](https://github.com/tlaplus/vscode-tlaplus/pull/307) (plus the modification I mentioned in #97 which is needed to build tlapm on my machine), to test out the VSCode extension. I've been having issues with proof steps that rely on Isabelle. For instance, the following proof obligation (where `seq` is defined as `<< >>`) that goes through in the most recent TLAPS distribution fails with false:

```
ASSUME NEW CONSTANT S,
       IsFiniteSet(S),
       1..Len(<<>>) = {}
       Isa
PROVE \E seq \in Seq(S) :
         /\ Len(seq) = 0
         /\ \A i, j \in 1..Len(seq) : i # j => seq[i] # seq[j]
```

Another type of failure I've observed is exemplified by the following:

```
ASSUME NEW CONSTANT ... (a number of constants not relevant to the obligation),
       NEW VARIABLE pc,
       NEW VARIABLE ... (variables in algvars)
       NEW VARIABLE arg,
       NEW VARIABLE ret,
       NEW VARIABLE P,
       P \in SUBSET ConfigDomain ,
       [Next]_algvars \o <<pc, arg, ret, P>> ,
       UNCHANGED (algvars \o <<pc, arg, ret, P>>) 
PROVE  P = P'
```

Here it is clearly the case that `UNCHANGED P` but this version of TLAPS is unable to infer this from `UNCHANGED (algvars \o <<pc, arg, ret, P>>)` – SMT times out. But this is not a problem in the most recent TLAPS distribution, where the obligation is easily resolved by SMT.

I'm not sure if these are known problems but I figured I'd report them here in case they weren't.

**Comment by kape1395** [comment]
Hi @uguryavuz, I had no problems with Isabelle in this branch (on Linux). The following theorem is checked via the LSP server. I had to move the `Isa` method from the assumptions to the BY clause. 
```
---- MODULE test_a ----
EXTENDS TLAPS, Naturals, FiniteSets

THEOREM
    ASSUME NEW CONSTANT S,
        IsFiniteSet(S),
        1..Len(<<>>) = {}
    PROVE \E seq \in Seq(S) :
            /\ Len(seq) = 0
            /\ \A i, j \in 1..Len(seq) : i # j => seq[i] # seq[j]
    BY Isa
====
```
The obligation is resolved by Isabelle (`proved isabelle[auto; time-limit: 30; time-used: 0.2 (1%)]`).
I don't have a Mac to check how does it run on it :/

**Comment by uguryavuz** [comment]
That fails immediately for me. I ran tlapm in command line and noticed the following line in the output:
```
Unknown logic "TLA+" -- no heap file found in:
  /Users/uguryavuz/.isabelle/Isabelle2011-1/heaps/polyml-undefined_unknown-platform
  /Users/uguryavuz/.opam/tlapm/lib/tlapm/backends/Isabelle/heaps/polyml-undefined_unknown-platform
```
which tells me that my workaround to get Isabelle to build causes some problems.

As an attempt to fix this, I merged the changes in [isabelle2020-dune](https://github.com/tlaplus/tlapm/tree/isabelle2020-dune) to this branch and built again. This did not work either – now I'm no longer observing that error but this instead:
```
### Missing Isabelle component: "/Users/uguryavuz/.opam/tlapm2020/lib/tlapm/backends/Isabelle/contrib/vscodium-1.70.1"
```

Any suggestions on how I could fix this?

**Comment by uguryavuz** [comment]
I gave up on trying to build `tlapm` and `tlapm_lsp` on my local machine for the time being and tried it out in a docker, but I'm still having some issues. I followed the instructions on a fresh Debian docker container and ran into no issues while building. `tlapm` builds without issues and I'm no longer having any issues with Isabelle. However, I don't seem to be able to run `tlapm-lsp`.

I can verify that it has been built successfully by running `opam exec -- tlapm_lsp --help` -- however I don't seem to be able to run the LSP itself:

```
u1@9cf03ae221a0:~/opt/tlapm$ opam exec -- tlapm_lsp --stdio
tlapm_lsp: internal error, uncaught exception:
           Unix.Unix_error(Unix.EINVAL, "io_uring_queue_init", "")
           Raised by primitive operation at Uring.create in file "lib/uring/uring.ml", line 398, characters 14-54
           Called from Eio_linux__Sched.with_sched in file "lib_eio_linux/sched.ml", line 534, characters 8-53
           Called from Dune__exe__Tlapm_lsp.Cli.term.combine in file "lsp/bin/tlapm_lsp.ml", line 90, characters 28-57
           Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
           Called from Cmdliner_term.ret.(fun) in file "cmdliner_term.ml", line 33, characters 25-32
           Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 34, characters 37-44
```

The exit status is 125.

File: Makefile
Comment (kape1395) [reply Makefile:47]: Thanks.

File: src/util/util.ml
Comment (kape1395) [reply src/util/util.ml:13]: Left these accidentally after experimenting with tree visitors.

File: src/util/util.mli
Comment (kape1395) [reply src/util/util.mli:13]: Left these accidentally after experimenting with tree visitors.

File: src/loader.mli
Comment (kape1395) [reply src/loader.mli:9]: Thanks. Replaced loaded by loader.

File: lsp/lib/docs/docs.mli
Comment (kape1395) [reply lsp/lib/docs/docs.mli:64]: Thanks.

File: lsp/lib/docs/proof_status.mli
Comment (kape1395) [reply lsp/lib/docs/proof_status.mli:15]: Thanks.

File: src/backend/interrupted.ml
Comment (kape1395) [reply src/backend/interrupted.ml:15]: @damiendoligez, maybe you know how to solve this?
Are there other ways to protect that state variable from concurrent access?

File: lsp/lib/docs/docs.mli
Comment (kape1395) [reply lsp/lib/docs/docs.mli:73]: Thanks. Refined the comment for get_proof_res.

File: lsp/lib/docs/obl.ml
Comment (kape1395) [reply lsp/lib/docs/obl.ml:68]: I was trying to make the errors more informative. But in this case `assert false` is probably ok. Changed it.

File: src/backend/interrupted.ml
Comment (kape1395) [reply src/backend/interrupted.ml:15]: Would the module Atomic work in this context?
https://github.com/ocaml/ocaml/blob/trunk/stdlib/atomic.mli

**Review by damiendoligez — COMMENTED** [review]
More comments.

File: .github/workflows/ocaml_versions.py
Comment (kape1395) [reply .github/workflows/ocaml_versions.py]: I merged the workflow changes from the main and discarded my local changes.

File: lsp/lib/docs/proof_step.ml
Comment (kape1395) [reply lsp/lib/docs/proof_step.ml:53]: Thanks.

File: lsp/lib/docs/proof_step.ml
Comment (kape1395) [reply lsp/lib/docs/proof_step.ml:25]: Thanks.

File: lsp/lib/docs/proof_step.mli
Comment (kape1395) [reply lsp/lib/docs/proof_step.mli:2]: Thanks.

File: lsp/lib/prover/progress.ml
Comment (kape1395) [reply lsp/lib/prover/progress.ml:10]: Thanks.

File: lsp/lib/prover/toolbox.ml
Comment (kape1395) [reply lsp/lib/prover/toolbox.ml:186]: Thanks.

File: lsp/lib/prover/toolbox.ml
Comment (kape1395) [reply lsp/lib/prover/toolbox.ml:148]: Thanks.

File: lsp/lib/prover/toolbox.ml
Comment (kape1395) [reply lsp/lib/prover/toolbox.ml:123]: Thanks.

File: lsp/lib/range.ml
Comment (kape1395) [reply lsp/lib/range.ml:117]: Thanks.

File: lsp/lib/prover/prover.ml
Comment (kape1395) [reply lsp/lib/prover/prover.ml:13]: Thanks.

File: lsp/lib/server/session.ml
Comment (kape1395) [reply lsp/lib/server/session.ml:62]: Thanks. Updated.

File: lsp/lib/server/session.ml
Comment (kape1395) [reply lsp/lib/server/session.ml:156]: Thanks. Simplified it.

File: lsp/lib/server/handlers.ml
Comment (kape1395) [reply lsp/lib/server/handlers.ml:39]: Thanks.

File: lsp/lib/server/handlers.ml
Comment (kape1395) [reply lsp/lib/server/handlers.ml:255]: Indeed. Thanks.

File: lsp/lib/server/handlers.ml
Comment (kape1395) [reply lsp/lib/server/handlers.ml:304]: Thanks.

File: lsp/lib/server/handlers.ml
Comment (kape1395) [reply lsp/lib/server/handlers.ml:335]: I tried to use `begin`/`end` before, but the standard OCaml formatter replaces them with parentheses.
I think it is better not to fight the formatter.

File: lsp/lib/prover/prover.ml
Comment (kape1395) [reply lsp/lib/prover/prover.ml:46]: Thanks.

File: lsp/lib/docs/proof_step.ml
Comment (kape1395) [reply lsp/lib/docs/proof_step.ml:27]: Indeed, description was messy.
Tried to reformulate it now.

File: lsp/lib/prover/toolbox.ml
Comment (kape1395) [reply lsp/lib/prover/toolbox.ml:54]: Thanks for pointing this out. I don't see how a client (LSP in this case) can determine if all attempts on the obligation are done. Can I extend the toolbox protocol on this by adding a list of pending provers in the initial message for the obligation? It would also be nice for the IDE user to see a list of pending provers for a step from the beginning. E.g.:

```
@!!type:obligation
@!!id:1
@!!loc:4:7:4:9
@!!status:to be proved
@!!provers:isabelle,z3 <------- added as a proposal.
@!!END
```

File: src/tlapm_lib.ml
Comment (damiendoligez) [inline src/tlapm_lib.ml:494]: This sentence needs to be updated.

File: src/tlapm_args.ml
Comment (damiendoligez) [inline src/tlapm_args.ml:62]: Is it still the case?

**Review by damiendoligez — APPROVED** [review]
This concludes my review. Good to merge after you look at these last few comments.

File: lsp/lib/prover/toolbox.ml
Comment (damiendoligez) [reply lsp/lib/prover/toolbox.ml:54]: That's a good idea, but we'll have to check that the toolbox itself properly ignores this new field.
@lemmy should be able to tell us.

File: lsp/lib/docs/proof_step.ml
Comment (damiendoligez) [inline lsp/lib/docs/proof_step.ml:28]: I think you mean "the BY keyword", rather than "the BY statement"?

File: lsp/lib/prover/toolbox.ml
Comment (lemmy) [reply lsp/lib/prover/toolbox.ml:54]: I wasn't involved in designing the TLAPS <> Toolbox protocol. I can start investigating, but I suggest we ask @xxyzzn at the next TLA+ Community meeting if there was a specific reason for not including this in the protocol originally.

File: lsp/lib/docs/proof_step.ml
Comment (kape1395) [reply lsp/lib/docs/proof_step.ml:28]: Indeed. Thanks.

File: src/tlapm_args.ml
Comment (kape1395) [reply src/tlapm_args.ml:274]: Good point. Changed it, thanks.

File: src/tlapm_args.ml
Comment (kape1395) [reply src/tlapm_args.ml:62]: Commit 47afb6b6c0fbfc297b9ffdbf8d73f943ddc11a99 resolves this.

File: src/tlapm_lib.ml
Comment (kape1395) [reply src/tlapm_lib.ml:494]: Tried to rephrase it:

> The flag `Params.use_stdin` can only be set to true if only a single file is passed to the `tlapm`. Therefore, the following assignment of the `use_stdin_prop` property will only be applied to a single file. This way, the file passed explicitly will be read from `stdin`, and all the files referenced from it will be searched in a file system, as usual.

File: src/loader.ml
Comment (kape1395) [reply src/loader.ml:89]: Resolved in 483732842b48953dfc56801fc42318ac254501ae by adding

```
  let try_read_file path =
    try
      let ic = open_in path in
      let content = In_channel.input_all ic in
      let () = close_in ic in
      Some content
    with Sys_error _ -> None
```

I decided to consider a failure to close the channel similarly to avoid unnecessary code complications.

File: lsp/lib/prover/toolbox.ml
Comment (kape1395) [reply lsp/lib/prover/toolbox.ml]: Resolved in 8138f4bc6ac1b65015bc82fb7871677b26dd467b.
A switch `--toolbox-vsn` was added, which defaults to `1` -- the old version.
In the new version (`2`) a message like the following is printed additionally:
```
@!!type:obligationprovers
@!!id:1
@!!provers:smt,zenon,isabelle
@!!END
```

On the LSP side I handle this message and check the finality of an obligation's
state without relying on the order of the provers.

**Comment by kape1395** [comment]
@damiendoligez, it seems I updated the PR according to all the comments. Thanks; they were all helpful.

**Comment by lemmy** [comment]
Minor regression: https://github.com/tlaplus/tlapm/pull/149

**Pull request merged**
